src/demangler/demangler.cc

```diff
@@ -0,0 +1,155 @@
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+#include "demangler/demangler.hpp"
+
+#include <string.h> // isspace, isxdigit, islower
+#include <unordered_map>
+
+#include <llvm/Demangle/Demangle.h>
+
+// With some exceptions we don't handle here, v0 Rust symbols can end in a
+// prefix followed by a 16-hexdigit hash, which must be removed
+constexpr std::string_view hash_pre = "::h";
+constexpr std::string_view hash_eg = "0123456789abcdef";
+
+// Minimal check that a string can end, and does end, in a hashlike substring
+// Some tools check for entropy, we do not.
+static inline bool has_hash(const std::string &str) {
+  // If the size can't conform, then the string is invalid
+  if (str.size() <= hash_pre.size() + hash_eg.size()) {
+    return false;
+  }
+
+  // Check that the string contains the hash prefix in the right position
+  if (str.compare(str.size() - hash_eg.size() - hash_pre.size(),
+                  hash_pre.size(), hash_pre)) {
+    return false;
+  }
+
+  // Check that the string ends in lowercase hex digits
+  for (size_t i = str.size() - hash_eg.size(); i < str.size(); ++i) {
+    if (!isxdigit(str[i]) || (!islower(str[i]) && !isdigit(str[i]))) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Rather than performing a strict check (which would be necessary for
+// supporting the use of '$' on some platforms as an informative token), this
+// implementation makes a minimal check indicating that a string is likely to
+// be a mangled Rust name.
+static bool is_probably_rust_legacy(const std::string &str) {
+  // Is the string too short to have a hash part in thefirst place?
+  if (!has_hash(str))
+    return false;
+
+  // Throw out `$$` and `$????$`, but not in-between
+  const char *ptr = str.c_str();
+  const char *end = ptr + str.size() - hash_pre.size() - hash_eg.size();
+  for (; ptr <= end; ++ptr) {
+    if (*ptr == '$') {
+      if (ptr[1] == '$') {
+        return false;
+      } else if (ptr[2] == '$' || ptr[3] == '$' || ptr[4] == '$') {
+        return true;
+      } else {
+        return false;
+      }
+    } else if (*ptr == '.') {
+      return !('.' == ptr[1] &&
+               '.' == ptr[2]); // '.' and '..' are fine, '...' is not
+    }
+  }
+  return true;
+}
+
+// Simple conversion from hex digit to integer
+// Includes capitals out of completeness, but this should not be necessary for
+// the implementation
+inline static int hex_to_int(char dig) {
+  if (dig >= '0' && dig <= '9')
+    return dig - '0';
+  else if (dig >= 'a' && dig <= 'f')
+    return dig - 'a' + 10;
+  else if (dig >= 'A' && dig <= 'F')
+    return dig - 'A' + 10;
+  else
+    return -1;
+}
+
+// Demangles a Rust string by building a copy piece-by-piece
+static std::string rust_demangle(const std::string &str) {
+  static const std::unordered_map<std::string, std::string> patterns{
```

Comment:
Why use an `std::unordered_map` ? 
`patterns` is never used as a map

Reply:
Yup, you're right.  Changed to a `std::array` of pairs in 8bb02a9.